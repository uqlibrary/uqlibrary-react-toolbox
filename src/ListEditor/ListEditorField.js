import React from 'react';
import ListEditor from './components/ListEditor';

export default function ListEditorField(fieldProps) {
    return(<ListEditor onChange={fieldProps.input.onChange} {...fieldProps} />);
}
