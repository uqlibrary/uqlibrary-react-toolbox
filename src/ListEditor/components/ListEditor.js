import React, {Component} from 'react';
import PropTypes from 'prop-types';
import ListRowHeader from './ListRowHeader';
import ListRow from './ListRow';
import ListForm from './ListForm';

export default class ListsEditor extends Component {

    static propTypes = {
        className: PropTypes.string,
        searchKey: PropTypes.object.isRequired,
        maxCount: PropTypes.number,
        isValid: PropTypes.func,
        disabled: PropTypes.bool,
        onChange: PropTypes.func,
        locale: PropTypes.object
    };

    static defaultProps = {
        maxCount: 0,
        searchKey: {
            value: 'rek_value',
            order: 'rek_order'
        }
    };

    constructor(props) {
        super(props);

        this.state = {
            itemList: []
        };
    }

    componentWillUpdate(nextProps, nextState) {
        // notify parent component when local state has been updated, eg itemList added/removed/reordered
        if (this.props.onChange) {
            this.props.onChange(this.transformOutput(nextState.itemList));
        }
    }

    transformOutput = (items) => {
        return items.map((item, index) => {
            return {
                [this.props.searchKey.value]: item,
                [this.props.searchKey.order]: index + 1
            };
        });
    }

    addItem = (item) => {
        if (this.props.maxCount === 0 || this.state.itemList.length < this.props.maxCount) {
            this.setState({
                itemList: [ ...this.state.itemList, item]
            });
        }
    }

    moveUpList = (item, index) => {
        if (index === 0) return;
        const nextList = this.state.itemList[index - 1];
        this.setState({
            itemList: [
                ...this.state.itemList.slice(0, index - 1),
                item, nextList,
                ...this.state.itemList.slice(index + 1)]
        });
    }

    moveDownList = (item, index) => {
        if (index === (this.state.itemList.length - 1)) return;
        const nextList = this.state.itemList[index + 1];
        this.setState({
            itemList: [
                ...this.state.itemList.slice(0, index),
                nextList, item,
                ...this.state.itemList.slice(index + 2)]
        });
    }

    deleteItem = (item, index) => {
        this.setState({
            itemList: this.state.itemList.filter((_, i) => i !== index),
        });
    }

    deleteAllItems = () => {
        this.setState({
            itemList: []
        });
    }

    render() {
        const renderListsRows = this.state.itemList.map((item, index) => (
            <ListRow
                key={index}
                index={index}
                item={item}
                canMoveDown={index !== this.state.itemList.length - 1}
                canMoveUp={index !== 0}
                onMoveUp={this.moveUpList}
                onMoveDown={this.moveDownList}
                onDelete={this.deleteItem}
                {...(this.props.locale && this.props.locale.row ? this.props.locale.row : {})}
                disabled={this.props.disabled} />
        ));

        return (
            <div className={this.props.className}>
                <ListForm
                    onAdd={this.addItem}
                    {...(this.props.locale && this.props.locale.form ? this.props.locale.form : {})}
                    isValid={this.props.isValid}
                    disabled={this.props.disabled || (this.props.maxCount > 0 && this.state.itemList.length >= this.props.maxCount)} />
                {
                    this.state.itemList.length > 0 &&
                    <ListRowHeader
                        {...(this.props.locale && this.props.locale.header ? this.props.locale.header : {})}
                        onDeleteAll={this.deleteAllItems}
                        disabled={this.props.disabled} />
                }
                {renderListsRows}
            </div>
        );
    }
}
