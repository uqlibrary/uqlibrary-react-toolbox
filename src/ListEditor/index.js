export {default as ListEditorField} from './ListEditorField';
