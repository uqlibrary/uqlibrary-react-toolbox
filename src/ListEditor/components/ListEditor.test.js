jest.dontMock('./ListEditor');

import ListEditor from './ListEditor';

function setup(testProps) {
    // props
    // formComponent: PropTypes.func.isRequired,
    // inputField: PropTypes.func, // eg connected auto complete fields
    // className: PropTypes.string,
    // searchKey: PropTypes.object.isRequired,
    // maxCount: PropTypes.number,
    // isValid: PropTypes.func,
    // disabled: PropTypes.bool,
    // onChange: PropTypes.func,
    // locale: PropTypes.object,
    // hideReorder: PropTypes.bool,
    // distinctOnly: PropTypes.bool,
    // errorText: PropTypes.string

    const props = {
        ...testProps,
        className: testProps.className || 'testClass', // : PropTypes.string,
        searchKey: testProps.searchKey || {value: 'value', order: 'order'}, // : PropTypes.object.isRequired,
        isValid: testProps.isValid || jest.fn(), // PropTypes.func,
        disabled: testProps.disabled || false, // PropTypes.bool,
        onChange: testProps.onChange || jest.fn(), // PropTypes.func,
        formComponent: testProps.formComponent || jest.fn(),
        inputField: testProps.inputField || jest.fn(),
        hideReorder: testProps.hideOrder || false,
        distinctOnly: testProps.distinctOnly || false,
        errorText: testProps.errorText || ''
    };

    return getElement(ListEditor, props);
}

describe('ListEditor tests ', () => {
    it('should render full component with a defined className', () => {
        const wrapper = setup({ className: 'requiredField' });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render full component as disabled', () => {
        const wrapper = setup({ disabled: true });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render an item to the list', () => {
        const wrapper = setup({ });
        expect(wrapper.state().itemList.length).toEqual(0);
        wrapper.instance().addItem('one');
        expect(wrapper.state().itemList.length).toEqual(1);
    });

    it('should render items not more than maxCount', () => {
        const wrapper = setup({ maxCount: 1 });
        expect(wrapper.state().itemList.length).toEqual(0);
        wrapper.instance().addItem('one');
        expect(wrapper.state().itemList.length).toEqual(1);
        wrapper.instance().addItem('two');
        expect(wrapper.state().itemList.length).toEqual(1);
    });

    it('should delete an item from the list', () => {
        const wrapper = setup({ });
        wrapper.setState({ itemList: [ 'one', 'two', 'three' ]});
        expect(wrapper.state().itemList.length).toEqual(3);
        wrapper.instance().deleteItem('one', 0);
        expect(wrapper.state().itemList.length).toEqual(2);
    });

    it('should delete all items from a list', () => {
        const wrapper = setup({ });
        wrapper.setState({ itemList: ['one', 'two', 'three'] });
        expect(wrapper.state().itemList.length).toEqual(3);
        wrapper.instance().deleteAllItems();
        expect(wrapper.state().itemList.length).toEqual(0);
    });

    it('should move up an item', () => {
        const wrapper = setup({ });
        wrapper.setState({ itemList: ['one', 'two', 'three']});
        expect(wrapper.state().itemList.length).toEqual(3);
        expect(wrapper.state().itemList[1]).toEqual('two');
        wrapper.instance().moveUpList('two', 1);
        expect(wrapper.state().itemList.length).toEqual(3);
        expect(wrapper.state().itemList[1]).toEqual('one');
    });

    it('should move down an item', () => {
        const wrapper = setup({ });
        wrapper.setState({ itemList: ['one', 'two', 'three']});
        expect(wrapper.state().itemList.length).toEqual(3);
        expect(wrapper.state().itemList[1]).toEqual('two');
        wrapper.instance().moveDownList('two', 1);
        expect(wrapper.state().itemList.length).toEqual(3);
        expect(wrapper.state().itemList[1]).toEqual('three');
    });
});
