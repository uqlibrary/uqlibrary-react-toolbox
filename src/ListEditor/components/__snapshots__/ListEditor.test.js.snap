// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEditor tests  should render full component as disabled 1`] = `
<div
  className="testClass"
>
  <mockConstructor
    disabled={true}
    errorText=""
    inputField={[MockFunction]}
    isValid={[MockFunction]}
    onAdd={[Function]}
  />
</div>
`;

exports[`ListEditor tests  should render full component with a defined className 1`] = `
<div
  className="requiredField"
>
  <mockConstructor
    disabled={false}
    errorText=""
    inputField={[MockFunction]}
    isValid={[MockFunction]}
    onAdd={[Function]}
  />
</div>
`;
